test/src/test/integration/link_git_protocol.rs

```diff
@@ -6,91 +6,106 @@
 use std::{
     collections::BTreeSet,
     io,
-    path::Path,
+    path::{Path, PathBuf},
     sync::{atomic::AtomicBool, Arc},
 };
 
 use bstr::ByteSlice as _;
 use futures::{AsyncReadExt as _, TryFutureExt as _};
+use git::{
+    prelude::*,
+    refs::transaction::{Change, PreviousValue, RefEdit},
+};
+use git_repository as git;
 use link_git_protocol::{fetch, ls, packwriter, upload_pack, ObjectId, PackWriter, Ref};
 use tempfile::{tempdir, TempDir};
 
 fn upstream() -> TempDir {
     let tmp = tempdir().unwrap();
 
-    let repo = git2::Repository::init_bare(&tmp).unwrap();
-    let auth = git2::Signature::now("apollo", "apollo@cree.de").unwrap();
-
-    let tree = {
-        let empty = repo.treebuilder(None).unwrap();
-        let oid = empty.write().unwrap();
-        repo.find_tree(oid).unwrap()
-    };
-    let base = {
-        let oid = repo
-            .commit(
-                Some("refs/namespaces/foo/refs/heads/main"),
-                &auth,
-                &auth,
-                "initial",
-                &tree,
-                &[],
-            )
-            .unwrap();
-        repo.find_commit(oid).unwrap()
+    let repo = git::init_bare(&tmp).unwrap().into_easy();
+    let auth = git::actor::Signature::now_local_or_utc("apollo", "apollo@cree.de");
+
+    let empty_tree_id = repo
+        .write_object(&git::objs::Tree::empty().into())
+        .unwrap()
+        .detach();
+    let base_id = {
+        repo.commit(
+            "refs/namespaces/foo/refs/heads/main",
+            "initial",
+            auth.clone(),
+            auth.clone(),
+            empty_tree_id,
+            git::commit::NO_PARENT_IDS,
+        )
+        .unwrap()
     };
-    let next = repo
+    let next_commit_id = repo
         .commit(
-            Some("refs/namespaces/foo/refs/heads/next"),
-            &auth,
-            &auth,
+            "refs/namespaces/foo/refs/heads/next",
```

Comment:
Now that I see the diff, there is probably no good reason to change the order of arguments here. If there is agreement that should be fixed.

Reply:
I don't mind, really, I can't remember 6 arguments anyway :)

No implicit `HEAD` is probably fair, but I'm surprised that the last 4 arguments are consumed.

Reply:
> I don't mind, really, I can't remember 6 arguments anyway :)

Ha, fair enough. It's not as bad as I made it sound, it's 'just' the message that moved underneath the author and committer fields. Keeping more similar would help with transitioning code, but then again, there is enough that's very different. I will just keep it like that then.

> No implicit HEAD is probably fair, but I'm surprised that the last 4 arguments are consumed.

I would be too, and there is a reason that hopefully actually is one in practice: APIs won't hide that they need to own an object by only asking for a reference. That way, if the caller has that object to spare, it's free. If the signature of a function asks for reference types, even if an owned version is passed in as `AsRef<oid>` or `AsRef<str>`, internally one has to go through a ref like `input.as_ref().to_owned()` even though the caller did something like `f(x)` and is happy that way. This adds a burden to callers as they have to clone if needed.

Reply:
Let me rephrase that: all arguments are `impl (Try)Into<Owned>`, which suggests that a copy occurs if a reference is passed. It might be that `git2` is lying, but it is still surprising that this cannot be done "zero copy", especially for object ids. So, why is that?

Reply:
> […] but it is still surprising that this cannot be done "zero copy", especially for object ids. So, why is that?

Amazing, I am glad you ask because I already took it for granted even though there isn't any other reason that I only implemented serialization for owned objects, not for (zero-copy) parsed ones. 

the `gixp pack-verify` program has a flag to re-encode all objects it sees to see if its rountrippable. Amazingly this is only about a tiny amount of time smaller than not doing that, so it might not be worth optimizing for this case.

However, it can definitely be done and I'd say it's not particularly hard. Those who have owned objects could then call `to_ref()` and get the immutable version that can then be serialized.

Sounds like it's old tech-debt that should probably be fixed once and for all. What do you think?

Reply:
Thanks for the explanation! 

It is probably not a bottleneck, but _maybe_ something we'd want to watch out for in the future. I just thought I'll ask, because during the making-of easy a lot of effort went into avoiding allocations for object ids.